tests/integration/test_bundle.py

```diff
@@ -23,15 +37,41 @@ def get_this_script_dir() -> Path:
     return Path(path)
 
 
+class State:
+    """Object used to hold the (incremental) test state."""
+
+    client_id: str
+    client_secret: str
+    redirect_uri: str
+
+
+@pytest.fixture(scope="module")
+def state() -> State:
+    return State()
+
+
 async def get_unit_address(ops_test: OpsTest, app_name: str, unit_num: int) -> str:
     """Get private address of a unit."""
     status = await ops_test.model.get_status()  # noqa: F821
     return status["applications"][app_name]["units"][f"{app_name}/{unit_num}"]["address"]
 
 
+async def get_app_address(ops_test: OpsTest, app_name: str) -> str:
+    """Get address of an app."""
+    status = await ops_test.model.get_status()  # noqa: F821
+    return status["applications"][app_name]["public-address"]
+
+
+async def get_reverse_proxy_app_url(
+    ops_test: OpsTest, ingress_app_name: str, app_name: str
+) -> str:
+    address = await get_app_address(ops_test, ingress_app_name)
+    return f"https://{address}/{ops_test.model.name}-{app_name}/"
+
+
 @pytest.mark.skip_if_deployed
 @pytest.mark.abort_on_fail
-async def test_render_and_deploy_bundle(ops_test: OpsTest):
+async def test_render_and_deploy_bundle(ops_test: OpsTest, dex: None) -> None:
```

Comment:
Nitpick:
- I guess it would be better to improve the naming `dex`.
- Not sure why the type is `NoneType`
- Have a concern that what if the external IDP service is changed from `dex` to some other one. I guess it would be better not to particularly rely on a specific implementation or service provider. I suggest we rely on some abstractions or interfaces, probably named `ext_idp_service` or something else.

Reply:
1) Fair point, I will change it.
2) Good catch